universum/github_handler.py

```diff
@@ -0,0 +1,61 @@
+import json
+import sys
+import requests
+
+from .modules.automation_server.jenkins_server import JenkinsServerForTrigger
+from .modules.vcs.github_vcs import GithubToken
+from .modules.output import needs_output
+from .modules.structure_handler import needs_structure
+from .lib.utils import make_block
+
+
+@needs_structure
+@needs_output
+class GithubHandler(JenkinsServerForTrigger, GithubToken):
+    """
+    Excepts payload from stdin!
+    """
+
+    @staticmethod
+    def define_arguments(argument_parser):
+        argument_parser.add_argument('--event', '-e', dest='event', metavar="GITHUB_EVENT",
+                                     help='Currently parsed from "x-github-event" header')
+        argument_parser.add_argument('--payload', '-p', dest='payload', metavar="PAYLOAD",
+                                     help='<actual help coming later> leave "-" to read from stdin')
+
+    def __init__(self, *args, **kwargs):
+        super(GithubHandler, self).__init__(*args, **kwargs)
+        if self.settings.payload == '-':
+            self.payload = json.loads(sys.stdin.read())
+        else:
+            self.payload = json.loads(self.settings.payload)
+
+    @make_block("Analysing trigger payload")
+    def execute(self):
+        if self.settings.event == "check_suite" and (self.payload["action"] in ["requested", "rerequested"]):
+            url = self.payload["repository"]["url"] + "/check-runs"
+            data = {"name": "CI tests", "head_sha": self.payload["check_suite"]["head_sha"]}
+            headers = {'Authorization': f"token {self.get_token(self.payload['installation']['id'])}",
+                       'Accept': 'application/vnd.github.antiope-preview+json'}
+            response = requests.post(url=url, json=data, headers=headers)
+            response.raise_for_status()
+
+        elif self.settings.event == "check_run" and \
+                (self.payload["action"] in ["requested", "rerequested", "created"]) and \
+                (str(self.payload["check_run"]["app"]["id"]) == str(self.settings.integration_id)):
+            self.trigger_build({
+                "GIT_REFSPEC": self.payload["check_run"]["check_suite"]["head_branch"],
```

Comment:
The loading of the JSON and using of the fields is placed to isolated parts of the code.
However, they are logically connected, so I prefer them to be inside the `execute` function.
Another issue: it will be hard to debug the issue with the format based on logs with this approach.
Probably, the parsing must be more resistant to the malformed data with proper error handling.